pkg/authn/identityheaders/identityheaders.go

```diff
@@ -52,7 +53,16 @@ func WithAuthHeaders(handler http.Handler, cfg *AuthnHeaderConfig) http.Handler
 			// Seemingly well-known headers to tell the upstream about user's identity
 			// so that the upstream can achieve the original goal of delegating RBAC authn/authz to kube-rbac-proxy
 			req.Header.Set(cfg.UserFieldName, u.GetName())
-			req.Header.Set(cfg.GroupsFieldName, strings.Join(u.GetGroups(), cfg.GroupSeparator))
+
+			filteredGroups := filterGroups(u.GetGroups(), cfg.GroupSeparator)
+
+			if cfg.GroupSeparator == "" {
+				for _, group := range filteredGroups {
+					req.Header.Add(cfg.GroupsFieldName, group)
+				}
+			} else {
+				req.Header.Set(cfg.GroupsFieldName, strings.Join(filteredGroups, cfg.GroupSeparator))
+			}
```

Comment:
nit: since we now know we don't have to filter groups when a separator is not set, I would just:
```golang
if cfg.GroupSeparator == "" {
	for _, group := range u.GetGroups() {
			req.Header.Add(cfg.GroupsFieldName, group)
	}
} else {
	filteredGroups := filterGroups(u.GetGroups(), cfg.GroupSeparator)
	req.Header.Set(cfg.GroupsFieldName, strings.Join(filteredGroups, cfg.GroupSeparator))
}
```

and remove the `if separator == ""` branch in `filterGroups()`